gps.1
'\" t
.\"     Title: gps
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 9 Aug 2004
.\"    Manual: GPSD Documentation
.\"    Source: The GPSD Project
.\"  Language: English
.\"
.TH "GPS" "1" "9 Aug 2004" "The GPSD Project" "GPSD Documentation"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gps, xgps, xgpsspeed, cgps, lcdgps, gegps \- test clients for gpsd
.SH "SYNOPSIS"
.HP \w'\fBxgps\fR\ 'u
\fBxgps\fR [\-D\ \fIdebug\-level\fR] [\-h] [\-V] [\-l\ [[d]\ |\ [m]\ |\ [s]]] [\-u\ [[i]\ |\ [n]\ |\ [m]]] [\fIserver\fR [\fI:port\fR [\fI:device\fR]]]
.HP \w'\fBxgpsspeed\fR\ 'u
\fBxgpsspeed\fR [\-D\ \fIdebug\-level\fR] [\-h] [\-V] [\-\-nautical] [\-\-landspeed] [\-\-speedunits\ {[mph]\ |\ [kph]\ |\ [knots]}] [\fIserver\fR [\fI:port\fR [\fI:device\fR]]]
.HP \w'\fBcgps\fR\ 'u
\fBcgps\fR [\-D\ \fIdebug\-level\fR] [\-h] [\-V] [\-l\ [[d]\ |\ [m]\ |\ [s]]] [\-m] [\-s] [\-u\ [[i]\ |\ [n]\ |\ [m]]] [\fIserver\fR [\fI:port\fR [\fI:device\fR]]]
.HP \w'\fBlcdgps\fR\ 'u
\fBlcdgps\fR [\-h] [\-V] [\-l\ [[d]\ |\ [m]\ |\ [s]]] [\-u\ [[i]\ |\ [n]\ |\ [m]]] [\fIserver\fR [\fI:port\fR [\fI:device\fR]]]
.HP \w'\fBgegps\fR\ 'u
\fBgegps\fR [\-d\ \fIdirectory\fR] [\-i]
.SH "DESCRIPTION"
.PP
These are the demonstration clients shipped with
gpsd\&. They have some common options:
.PP
The
\fB\-h\fR
option causes each client to emit a summary of its options and then exit\&.
.PP
The
\fB\-V\fR
option causes each client to dump the package version and exit\&.
.PP
The
\fB\-l\fR
option, when present, sets the format of latitude and longitude reports\&. The value \*(Aqd\*(Aq produces decimal degrees and is the default\&. The value \*(Aqm\*(Aq produces degrees and decimal minutes\&. The value \*(Aqs\*(Aq produces degrees, minutes, and decimal seconds\&.
.PP
xgps,
cgps, and
lcdgps
look at variables in the environment to figure out what units they should default to using for display \(em imperial, nautical, or metric\&. Here are the variables and values they check:
.sp
.if n \{\
.RS 4
.\}
.nf
    GPSD_UNITS one of:
              imperial   = miles/feet
              nautical   = knots/feet
              metric     = km/meters
    LC_MEASUREMENT
	      en_US      = miles/feet
              C          = miles/feet
              POSIX      = miles/feet
              [other]    = km/meters
    LANG
	      en_US      = miles/feet
              C          = miles/feet
              POSIX      = miles/feet
              [other]    = km/meters
.fi
.if n \{\
.RE
.\}
.PP
These preferences may be overridden by the
\fB\-u\fR
option\&.
.PP
Where present, the
\fB\-u\fR
option can be used to set the system units for display; follow the keyword with \*(Aqi\*(Aq for \*(Aqimperial\*(Aq for American units (feet in altitude and error estimates, miles per hour in speeds), \*(Aqn\*(Aq for \*(Aqnautical\*(Aq (feet in altitude and error estimates, knots in speed) or \*(Aqm\*(Aq for \*(Aqmetric\*(Aq (meters in altitude and error estimates, kilometers per hour in speeds)\&.
.PP
The
\fB\-D\fR
option, when present, sets a debug level; it is primarily for use by GPSD developers\&. It enables various progress messages to standard error\&.
.PP
By default, clients collect data from all compatible devices on localhost, using the default GPSD port 2947\&. An optional argument to any client may specify a server to get data from\&. A colon\-separated suffix is taken as a port number\&. If there is a second colon\-separated suffix, that is taken as a specific device name to be watched\&. However, if the server specification contains square brackets, the part inside them is taken as an IPv6 address and port/device suffixes are only parsed after the trailing bracket\&. Possible cases look like this:
.PP
localhost:/dev/ttyS1
.RS 4
Look at the default port of localhost, trying both IPv4 and IPv6 and watching output from serial device 1\&.
.RE
.PP
example\&.com:2317
.RS 4
Look at port 2317 on example\&.com, trying both IPv4 and IPv6\&.
.RE
.PP
71\&.162\&.241\&.5:2317:/dev/ttyS3
.RS 4
Look at port 2317 at the specified IPv4 address, collecting data from attached serial device 3\&.
.RE
.PP
[FEDC:BA98:7654:3210:FEDC:BA98:7654:3210]:2317:/dev/ttyS5
.RS 4
Look at port 2317 at the specified IPv6 address, collecting data from attached serial device 5\&.
.RE
.PP
Not all clients shipped with GPSD are documented here\&. See also the separate manual pages for
\fBgpspipe\fR(1)
and
\fBgpsmon\fR(1)
and
\fBgpxlogger\fR(1)
\&.
.SS "xgps"
.PP
xgps
is a simple test client for
gpsd
with an X interface\&. It displays current GPS position/time/velocity information and (for GPSes that support the feature) the locations of accessible satellites\&.
.PP
In the sky view, satellites are color\-coded to indicate quality of signal; consult the data display to the left for exact figures in dB\&. Square icons indicate WAAS/EGNOS satellites, circles indicate ordinary GPS satellites\&. Filled icons were used in the last fix, outline icons were not\&.
.SS "xgpsspeed"
.PP
xgpsspeed
is a speedometer that uses position information from the GPS\&. It accepts an \-h option and optional argument as for
gps, or a \-V option to dump the package version and exit\&.
.PP
The default display mode is a speed and track presentation modeled after a marine navigation display; for backward compatibility the \-\-nautical option forces this mode\&. The \-\-landspeed option produces a simple speedometer\&.
.PP
The \-speedunits option can be used to set the speed units for display; follow the keyword with knots for nautical miles per hour, kph for kilometres per hour, or mph for miles per hour\&. The default is miles per hour\&.
.PP
In the nautical mode only, \-\-maxspeed sets the maximum on the speedometer\&.
.SS "cgps"
.PP
cgps
is a client resembling
xgps, but without the pictorial satellite display and able to run on a serial terminal or terminal emulator\&.
.PP
The
\fB\-s\fR
option prevents
cgps
from displaying the data coming from the daemon\&. This display can also be toggled with the s command\&.
.PP
The
\fB\-m\fR
option will display your magnetic heading (as opposed to your true heading)\&. This is a calculated value, not a measured value, and is subject to a potential error of up to two degrees in the areas for which the calculation is valid (currently Western Europe, Alaska, and Lower 48 in the USA)\&. The formulas used are those found in the Aviation Formulary v1\&.43\&.
.PP
cgps
terminates when you send it a SIGHUP or SIGINT; given default terminal settings this will happen when you type Ctrl\-C at it\&. It will also terminate on \*(Aqq\*(Aq
.SS "lcdgps"
.PP
A client that passes
gpsd
data to
lcdproc, turning your car computer into a very expensive and nearly feature\-free GPS receiver\&. Currently assumes a 4x40 LCD and writes data formatted to fit that size screen\&. Also displays 4\- or 6\-character Maidenhead grid square output\&.
.SS "gegps"
.PP
This program collects fixes from
gpsd
and feeds them to a running instance of Google Earth for live location tracking\&.
.PP
The
\fB\-d\fR
argument is the location of the Google Earth installation directory\&. If not specified, it defaults to the current directory\&.
.PP
If you have the free (non\-subscription) version, start by running with the
\fB\-i\fR
option to drop a clue in the Google Earth installation directory, as \*(AqOpen_in_Google_Earth_RT_GPS\&.kml\*(Aq, then open that file in Places (File > Open\&.\&.\&.)\&. Run
gpsd
in the normal way after that\&.
.SH "SEE ALSO"
.PP
\fBgpsd\fR(8),
\fBlibgps\fR(3),
\fBlibgpsmm\fR(3),
\fBgpsfake\fR(1),
\fBgpsctl\fR(1),
\fBgpscat\fR(1),
\fBgpsprof\fR(1)\&.
\fBgpspipe\fR(1)\&.
\fBgpsmon\fR(1)\&.
\fBgpxlogger\fR(1)\&.
.SH "AUTHORS"
.PP
Remco Treffcorn, Derrick Brashear, Russ Nelson & Eric S\&. Raymond, Jeff Francis (cgps), Chen Wei
<weichen302@aol\&.com>
(gegps & xgpsspeed), Robin Wittler
<real@the\-real\&.org>
(xgpsspeed)\&.
.PP
This manual page by Eric S\&. Raymond
<esr@thyrsus\&.com>


debian/changelog
gpsd (3.17-3) unstable; urgency=medium

  * [2638b3e] Avoid duplicate ppscheck installation.
    Thanks to Andreas Beckmann (Closes: #879951)

 -- Bernd Zeimetz <bzed@debian.org>  Fri, 27 Oct 2017 17:26:55 +0200

gpsd (3.17-2) unstable; urgency=medium

  * Uploading to unstable.
  * [a13df427] Fixing last changelog entry.
    Some lines went missing, oops.

 -- Bernd Zeimetz <bzed@debian.org>  Thu, 26 Oct 2017 23:16:41 +0200

gpsd (3.17-1) experimental; urgency=medium

  * [d21fbf15] Updated version 3.17 from 'upstream/3.17'
    with Debian dir 3099dc33257e902647917f2ad7e67eec078b0b42
  * [913ea95c] Fix SConstruct.
  * [d9d7b300] Python dbg _d.so files are not built anymore.
  * [f0ea0948] Install ppscheck + manpage
  * [7df3aa0a] Updating symbols files

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 21 Oct 2017 16:12:28 +0200

gpsd (3.16-4) unstable; urgency=medium

  * [2405f456] Fix debug symbol handling.
    Debug symbols went missing when upstream switched the strip
    option name.
    Thanks to Andre Naujoks (Closes: #842792)
  * [74c3cd07] Updating debian/control from debian/control.in
  * [b66410da] hadle debian/control creation a bit better
  * [06a9623f] Use gz compression for now.
  * [44376877] Remove obsolete build dependency on hardening-includes.

 -- Bernd Zeimetz <bzed@debian.org>  Fri, 04 Nov 2016 23:35:14 +0100

gpsd (3.16-3) unstable; urgency=medium

  * [aa43cc65] Various patches were applied upstream.
    Take them from there, removing our version.
  * [1444f940] Build with -std=gnu++98. (Closes: #830993)
  * [505d032a] Remove netbase/systemd-sysv dependency.
    Not needed anymore. (Closes: #768100)
  * [b53473f0] Migrate to github.
  * [cb624dcb] Add travis.yml build stuff.
  * [1c3c3463] Updating gbp.conf to ignore travis.yml for debuild.
  * [d870a5c6] Don't rebuild for the static library.

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 24 Sep 2016 14:34:07 +0200

gpsd (3.16-2) unstable; urgency=medium

  * [5001c11a] Do not start gpsd.service by default.
    We use socket based activation only.
    If one needs a running gpsd.service at boot - without being started by
    gpsd.socket - you'll have to configure systemd to start it.
    Too many people have gpsd installed as dependency and do not need yet
    another running process.

 -- Bernd Zeimetz <bzed@debian.org>  Sun, 15 May 2016 15:35:20 +0200

gpsd (3.16-1) unstable; urgency=medium

  * [c3d16637] Merge tag 'upstream/3.16'
    Upstream version 3.16
  * [29863f7a] Snapshot changelog.
  * [e61a749f] refreshing patches.
  * [ddc25e61] Pick some patches from upstream.
    - Fix test_gpsmm building
    - Fix FTBFS on hurd (Closes: #821804)
  * [67d9aca8] Fix check for clock_gettime.
  * [67b39606] Update git-buildpackage config.

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 30 Apr 2016 20:07:13 +0200

gpsd (3.15-2) unstable; urgency=medium

  * Uploading to unstable

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 31 Oct 2015 20:02:31 +0100

gpsd (3.15-1) experimental; urgency=medium

  * [10e7caf] Install contrib/* as examples.
    Thanks to Sanjeev Gupta (Closes: #781688)
  * [2d150e4] Merge tag 'upstream/3.15'
    Upstream version 3.15
  * [4163f08] snapshot changelog.
  * [1ea8da4] Remove libgpsd.
    Upstream links it statically now and I can't be bothered to maintain
    patches as Gentoo and others do it.
    The result is a slightly larger gpsd-clients package and running
    the binaries in parallel will waste memory.
    Thanks to Paul Wise & Andreas Beckmann (Closes: #779286, #781567)
  * [bb7a556] scons distclean doesn't work anymore.
    use --clean + sconsclean.
  * [31ff968] Refreshing patches
  * [5861e60] Fix broken static linking of test_gpsmm
  * [ef84468] Install new files.
  * [f9e250c] Add patch to fix regression tests for leap second bump
  * [65b8ac6] Updating symbols file.
  * [7e3d4a4] Update libqgpsmm symbols.
  * [785e932] _STRUCT_TIMESPEC is not defined in time.h anymore.
    Only linux/time.h defines _STRUCT_TIMESPEC.
    Check for __timespec_defined additionally.
  * [0b56700] Fix struct member names in timspec.
  * [e723fdd] Use system clock_gettime again.
    Fix some more broken header checks in SConstruct.

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 31 Oct 2015 20:02:07 +0100

gpsd (3.12-1) experimental; urgency=medium

  * [9f0dfbda] Merge tag 'upstream/3.12'
    Upstream version 3.12
  * [0896c861] Refresh/update patches.
  * [9daccea4] Fix test_ debug tool list.
  * [85f4f0bb] Add patch to fix test_gpsmm.
  * [9c511016] Add some new symbols to the symbols file.
  * [30cf1b92] Add dh-python as build dependency.
  * [24c325f9] Bump library sonames in package names.
  * [a810ac32] Update libqgpsmm symbols file
  * [1bf373c6] Fix some lintian errors.

 -- Bernd Zeimetz <bzed@debian.org>  Mon, 23 Feb 2015 22:37:42 +0100

gpsd (3.11-3) unstable; urgency=medium

  [ Andreas Beckmann ]
  * [dd6bbdf1] avoid prompting due to modified lenny conffiles

 -- Bernd Zeimetz <bzed@debian.org>  Sun, 25 Jan 2015 12:15:44 +0100

gpsd (3.11-2) unstable; urgency=medium

  * [a759614c] Avoid prompting due to modified squeeze conffiles.
    Thanks to Andreas Beckmann (Closes: #774872)

 -- Bernd Zeimetz <bzed@debian.org>  Thu, 08 Jan 2015 21:35:30 +0100

gpsd (3.11-1) unstable; urgency=medium

  * [325d1900] Merge tag 'upstream/3.11'
    Upstream version 3.11
  * [8d86b938] Refreshing patches.
  * [877cbcf3] Update symbols file.


gps.py
#!/usr/bin/env python
# $Id: gps.py 4988 2009-01-12 17:11:54Z esr $
#
# gps.py -- Python interface to GPSD.
#
import time, calendar, math, socket, sys, select

NaN = float('nan')
def isnan(x): return str(x) == 'nan'

ONLINE_SET	= 0x00000001
TIME_SET	= 0x00000002
TIMERR_SET	= 0x00000004
LATLON_SET	= 0x00000008
ALTITUDE_SET	= 0x00000010
SPEED_SET	= 0x00000020
TRACK_SET	= 0x00000040
CLIMB_SET	= 0x00000080
STATUS_SET	= 0x00000100
MODE_SET	= 0x00000200
HDOP_SET  	= 0x00000400
VDOP_SET  	= 0x00000800
PDOP_SET  	= 0x00001000
TDOP_SET	= 0x00002000
GDOP_SET	= 0x00004000
HERR_SET	= 0x00008000
VERR_SET	= 0x00010000
PERR_SET	= 0x00020000
SATELLITE_SET	= 0x00040000
SPEEDERR_SET	= 0x00080000
TRACKERR_SET	= 0x00100000
CLIMBERR_SET	= 0x00200000
DEVICE_SET	= 0x00400000
DEVICELIST_SET	= 0x00800000
DEVICEID_SET	= 0x01000000
ERROR_SET	= 0x02000000
CYCLE_START_SET	= 0x04000000
FIX_SET		= (TIME_SET|MODE_SET|TIMERR_SET|LATLON_SET|HERR_SET|ALTITUDE_SET|VERR_SET|TRACK_SET|TRACKERR_SET|SPEED_SET|SPEEDERR_SET|CLIMB_SET|CLIMBERR_SET)

STATUS_NO_FIX = 0
STATUS_FIX = 1
STATUS_DGPS_FIX = 2
MODE_NO_FIX = 1
MODE_2D = 2
MODE_3D = 3
MAXCHANNELS = 12
SIGNAL_STRENGTH_UNKNOWN = NaN

# The spec says 82, but some receivers (TN-200, GSW 2.3.2) output 86 characters
NMEA_MAX = 86

GPSD_PORT = 2947

class gpstimings:
    def __init__(self):
        self.sentence_tag = ""
        self.sentence_length = 0
        self.sentence_time = 0.0
        self.d_xmit_time = 0.0
        self.d_recv_time = 0.0
        self.d_decode_time = 0.0
        self.emit_time = 0.0
        self.poll_time = 0.0
        self.c_recv_time = 0.0
        self.c_decode_time = 0.0
    def d_received(self):
        if self.sentence_time:
            return self.d_recv_time + self.sentence_time
        else:
            return self.d_recv_time + self.d_xmit_time
    def collect(self, tag, length, sentence_time, xmit_time, recv_time, decode_time, poll_time, emit_time):
        self.sentence_tag = tag
        self.sentence_length = int(length)
        self.sentence_time = float(sentence_time)
        self.d_xmit_time = float(xmit_time)
        self.d_recv_time = float(recv_time)
        self.d_decode_time = float(decode_time)
        self.poll_time = float(poll_time)
        self.emit_time = float(emit_time)
    def __str__(self):
        return "%s\t%2d\t%2.6f\t%2.6f\t%2.6f\t%2.6f\t%2.6f\t%2.6f\t%2.6f\t%2.6f\n" \
               % (self.sentence_tag,
                 self.sentence_length,
                 self.sentence_time,
                 self.d_xmit_time, 
                 self.d_recv_time,
                 self.d_decode_time,
                 self.poll_time,
                 self.emit_time,
                 self.c_recv_time,
                 self.c_decode_time)
        

class gpsfix:
    def __init__(self):
        self.mode = MODE_NO_FIX
        self.time = NaN
        self.ept = NaN
        self.latitude = self.longitude = 0.0
        self.eph = NaN
        self.altitude = NaN			# Meters
        self.epv = NaN
        self.track = NaN			# Degrees from true north
        self.speed = NaN			# Knots
        self.climb = NaN			# Meters per second
        self.epd = NaN
        self.eps = NaN
        self.epc = NaN

class gpsdata:
    "Position, track, velocity and status information returned by a GPS."

    class satellite:
        def __init__(self, PRN, elevation, azimuth, ss, used=None):
            self.PRN = PRN
            self.elevation = elevation
            self.azimuth = azimuth
            self.ss = ss
            self.used = used
        def __repr__(self):
            return "PRN: %3d  E: %3d  Az: %3d  Ss: %d Used: %s" % (self.PRN,self.elevation,self.azimuth,self.ss,"ny"[self.used])

    def __init__(self):
        # Initialize all data members 
        self.online = 0			# NZ if GPS on, zero if not

        self.valid = 0
        self.fix = gpsfix()

        self.status = STATUS_NO_FIX
        self.utc = ""

        self.satellites_used = 0		# Satellites used in last fix
        self.pdop = self.hdop = self.vdop = self.tdop = self.gdop = 0.0

        self.epe = 0.0

        self.satellites = []			# satellite objects in view
        self.await = self.parts = 0

        self.gps_id = None
        self.driver_mode = 0
        self.profiling = False
        self.timings = gpstimings()
        self.baudrate = 0
        self.stopbits = 0
        self.cycle = 0
        self.mincycle = 0
        self.device = None
        self.devices = []

    def __repr__(self):
        st = "Time: %s (%s)" % (self.utc, self.fix.time)
        st += "Lat/lon:  %f %f\n" % (self.fix.latitude, self.fix.longitude)
        if isnan(self.fix.altitude):
            st += "Altitude: ?\n"
        else:
            st += "Altitude: %f\n" % (self.fix.altitude)
        if isnan(self.fix.speed):
            st += "Speed:    ?\n"
        else:
            st += "Speed:    %f\n" % (self.fix.speed)
        if isnan(self.fix.track):
            st += "Track:    ?\n"
        else:
            st += "Track:    %f\n" % (self.fix.track)
        st += "Status:   STATUS_%s\n" %("NO_FIX","FIX","DGPS_FIX")[self.status]
        st += "Mode:     MODE_"+("ZERO", "NO_FIX", "2D","3D")[self.fix.mode]+"\n"
        st += "Quality:  %d p=%2.2f h=%2.2f v=%2.2f t=%2.2f g=%2.2f\n" % \
              (self.satellites_used, self.pdop, self.hdop, self.vdop, self.tdop, self.gdop)
        st += "Y: %s satellites in view:\n" % len(self.satellites)
        for sat in self.satellites:
          st += "    " + repr(sat) + "\n"
        return st

class gps(gpsdata):
    "Client interface to a running gpsd instance."
    def __init__(self, host="127.0.0.1", port="2947", verbose=0):
        gpsdata.__init__(self)
        self.sock = None	# in case we blow up in connect
        self.sockfile = None
        self.connect(host, port)
        self.verbose = verbose
        self.raw_hook = None

    def connect(self, host, port):
        """Connect to a host on a given port.

        If the hostname ends with a colon (`:') followed by a number, and
        there is no port specified, that suffix will be stripped off and the
        number interpreted as the port number to use.
        """
        if not port and (host.find(':') == host.rfind(':')):
            i = host.rfind(':')
            if i >= 0:
                host, port = host[:i], host[i+1:]
            try: port = int(port)
            except ValueError:
                raise socket.error, "nonnumeric port"
        if not port: port = GPSD_PORT
        #if self.debuglevel > 0: print 'connect:', (host, port)
        msg = "getaddrinfo returns an empty list"
        self.sock = None
        self.sockfile = None
        for res in socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM):
            af, socktype, proto, canonname, sa = res
            try:
                self.sock = socket.socket(af, socktype, proto)
                #if self.debuglevel > 0: print 'connect:', (host, port)
                self.sock.connect(sa)
                self.sockfile = self.sock.makefile()
            except socket.error, msg:
                #if self.debuglevel > 0: print 'connect fail:', (host, port)
                self.close()
                continue
            break
        if not self.sock:
            raise socket.error, msg

    def set_raw_hook(self, hook):
        self.raw_hook = hook

    def close(self):
        if self.sockfile:
            self.sockfile.close()
        if self.sock:
            self.sock.close()
        self.sock = None
        self.sockfile = None

    def __del__(self):
        self.close()

    def __unpack(self, buf):
        # unpack a daemon response into the instance members
        self.fix.time = 0.0
        fields = buf.strip().split(",")
        if fields[0] == "GPSD":
            for field in fields[1:]:
                if not field or field[1] != '=':
                  continue
                cmd = field[0]
                data = field[2:]
                if data[0] == "?":
                  continue
                if cmd in ('A', 'a'):
                  self.fix.altitude = float(data)
                  self.valid |= ALTITUDE_SET
                elif cmd in ('B', 'b'):
                  if data == '?':
                    self.baudrate = self.stopbits = 0
                    self.device = None
                  else:
                    (f1, f2, f3, f4) = data.split()
                    self.baudrate = int(f1)
                    self.stopbits = int(f4)
                elif cmd in ('C', 'c'):
                  if data == '?':
                    self.cycle = -1
                    self.device = None
                  elif len(data.split()) == 2:
                    (self.cycle, self.mincycle) = map(float, data)
                  else:
                    self.mincycle = self.cycle = float(data)
                elif cmd in ('D', 'd'):
                  self.utc = data
                  self.fix.time = isotime(self.utc)
                  self.valid |= TIME_SET
                elif cmd in ('E', 'e'):
                  parts = data.split()
                  (self.epe, self.fix.eph, self.fix.epv) = map(float, parts)
                  self.valid |= HERR_SET | VERR_SET | PERR_SET
                elif cmd in ('F', 'f'):
                  if data == '?':
                      self.device = None
                  else:
                      self.device = data
                elif cmd in ('I', 'i'):
                  if data == '?':
                    self.cycle = -1
                    self.gps_id = None
                  else:
                    self.gps_id = data
                elif cmd in ('K', 'K'):
                  if data == '?':
                      self.devices = None
                  else:
                      self.devices = data[1:].split()
                elif cmd in ('M', 'm'):
                  self.fix.mode = int(data)
                  self.valid |= MODE_SET
                elif cmd in ('N', 'n'):
                  if data == '?':
                    self.driver_mode = -1
                    self.device = None
                  else:
                    self.driver_mode = int(data)
                elif cmd in ('O', 'o'):
                  fields = data.split()
                  if fields[0] == '?':
                      self.fix.mode = MODE_NO_FIX
                  else:
                      self.timings.sentence_tag = fields[0]
                      def default(i, cnv=float):
                        if fields[i] == '?':
                            return NaN
                        else:
                            return cnv(fields[i])
                        self.utc = fields[1]
                      self.fix.time = default(1)
                      if not isnan(self.fix.time):
                        self.utc = isotime(self.fix.time)
                      self.fix.ept = default(2)
                      self.fix.latitude = default(3)
                      self.fix.longitude = default(4)
                      self.fix.altitude = default(5)
                      self.fix.eph = default(6)
                      self.fix.epv = default(7)
                      self.fix.track = default(8)
                      self.fix.speed = default(9)
                      self.fix.climb = default(10)
                      self.fix.epd = default(11)
                      self.fix.eps = default(12)
                      self.fix.epc = default(13)
                      if len(fields) > 14:
                        self.fix.mode = default(14, int)
                      else:
                        if isnan(self.fix.altitude):
                          self.fix.mode = MODE_2D
                        else:
                           self.fix.mode = MODE_3D
                      self.valid = TIME_SET|TIMERR_SET|LATLON_SET|MODE_SET
                      if self.fix.mode == MODE_3D:
                        self.valid |= ALTITUDE_SET | CLIMB_SET
                      if not isnan(self.fix.eph):
                        self.valid |= HERR_SET
                      if not isnan(self.fix.epv):
                        self.valid |= VERR_SET
                      if not isnan(self.fix.track):
                        self.valid |= TRACK_SET | SPEED_SET
                      if not isnan(self.fix.eps):
                        self.valid |= SPEEDERR_SET
                      if not isnan(self.fix.epc):
                        self.valid |= CLIMBERR_SET
                elif cmd in ('P', 'p'):
                  (self.fix.latitude, self.fix.longitude) = map(float, data.split())
                  self.valid |= LATLON_SET
                elif cmd in ('Q', 'q'):
                  parts = data.split()
                  self.satellites_used = int(parts[0])
                  (self.pdop, self.hdop, self.vdop, self.tdop, self.gdop) = map(float, parts[1:])
                  self.valid |= HDOP_SET | VDOP_SET | PDOP_SET | TDOP_SET | GDOP_SET
                elif cmd in ('S', 's'):
                  self.status = int(data)
                  self.valid |= STATUS_SET
                elif cmd in ('T', 't'):
                  self.fix.track = float(data)
                  self.valid |= TRACK_SET
                elif cmd in ('U', 'u'):
                  self.fix.climb = float(data)
                  self.valid |= CLIMB_SET
                elif cmd in ('V', 'v'):
                  self.fix.speed = float(data)
                  self.valid |= SPEED_SET
                elif cmd in ('X', 'x'):
                  if data == '?':
                      self.online = -1
                      self.device = None
                  else:
                      self.online = float(data)
                      self.valid |= ONLINE_SET
                elif cmd in ('Y', 'y'):
                    satellites = data.split(":")
                    prefix = satellites.pop(0).split()
                    self.timings.sentence_tag = prefix.pop(0)
                    self.timings.sentence_time = prefix.pop(0)
                    if self.timings.sentence_time != "?":
                        self.timings.sentence_time = float(self.timings.sentence_time)
                    d1 = int(prefix.pop(0))
                    newsats = []
                    for i in range(d1):
                        newsats.append(gps.satellite(*map(int, satellites[i].split())))
                    self.satellites = newsats
                    self.valid |= SATELLITE_SET
                elif cmd in ('Z', 'z'):
                  self.profiling = (data[0] == '1')
                elif cmd == '$':
                  self.timings.collect(*data.split())
        if self.raw_hook:
            self.raw_hook(buf);

    def waiting(self):
        "Return True if data is ready for the client."
        (winput,woutput,wexceptions) = select.select((self.sock,), (),(), 0)
        return winput != []

    def poll(self):
        "Wait for and read data being streamed from gpsd."
        self.response = self.sockfile.readline()
        if self.response.startswith("H") and "=" not in self.response:
            while True:
                frag = self.sockfile.readline()
                self.response += frag
                if frag.startswith("."):
                    break
        if not self.response:
            return -1
        if self.verbose:
            sys.stderr.write("GPS DATA %s\n" % repr(self.response))
        self.timings.c_recv_time = time.time()
        self.__unpack(self.response)
        if self.profiling:
            if self.timings.sentence_time != '?':
                basetime = self.timings.sentence_time
            else:
                basetime = self.timings.d_xmit_time
            self.timings.c_decode_time = time.time() - basetime
            self.timings.c_recv_time -= basetime
        return 0

    def send(self, commands):
        "Ship commands to the daemon."
        if not commands.endswith("\n"):
            commands += "\n"
        self.sock.send(commands)

    def query(self, commands):
        "Send a command, get back a response."
        self.send(commands)
        return self.poll()

# some multipliers for interpreting GPS output
METERS_TO_FEET	= 3.2808399
METERS_TO_MILES	= 0.00062137119
KNOTS_TO_MPH	= 1.1507794

# EarthDistance code swiped from Kismet and corrected
# (As yet, this stuff is not in the libgps C library.)

def Deg2Rad(x):
    "Degrees to radians."
    return x * (math.pi/180)

def Rad2Deg(x):
    "Radians to degress."
    return x * (180/math.pi)

def CalcRad(lat):
    "Radius of curvature in meters at specified latitude."
    a = 6378.137
    e2 = 0.081082 * 0.081082
    # the radius of curvature of an ellipsoidal Earth in the plane of a
    # meridian of latitude is given by
    #
    # R' = a * (1 - e^2) / (1 - e^2 * (sin(lat))^2)^(3/2)
    #
    # where a is the equatorial radius,
    # b is the polar radius, and
    # e is the eccentricity of the ellipsoid = sqrt(1 - b^2/a^2)
    #
    # a = 6378 km (3963 mi) Equatorial radius (surface to center distance)
    # b = 6356.752 km (3950 mi) Polar radius (surface to center distance)
    # e = 0.081082 Eccentricity
    sc = math.sin(Deg2Rad(lat))
    x = a * (1.0 - e2)
    z = 1.0 - e2 * sc * sc
    y = pow(z, 1.5)
    r = x / y

    r = r * 1000.0	# Convert to meters
    return r

def EarthDistance((lat1, lon1), (lat2, lon2)):
    "Distance in meters between two points specified in degrees."
    x1 = CalcRad(lat1) * math.cos(Deg2Rad(lon1)) * math.sin(Deg2Rad(90-lat1))
    x2 = CalcRad(lat2) * math.cos(Deg2Rad(lon2)) * math.sin(Deg2Rad(90-lat2))
    y1 = CalcRad(lat1) * math.sin(Deg2Rad(lon1)) * math.sin(Deg2Rad(90-lat1))
    y2 = CalcRad(lat2) * math.sin(Deg2Rad(lon2)) * math.sin(Deg2Rad(90-lat2))
    z1 = CalcRad(lat1) * math.cos(Deg2Rad(90-lat1))
    z2 = CalcRad(lat2) * math.cos(Deg2Rad(90-lat2))
    a = (x1*x2 + y1*y2 + z1*z2)/pow(CalcRad((lat1+lat2)/2), 2)
    # a should be in [1, -1] but can sometimes fall outside it by
    # a very small amount due to rounding errors in the preceding
    # calculations (this is prone to happen when the argument points
    # are very close together).  Thus we constrain it here.
    if abs(a) > 1: a = 1
    elif a < -1: a = -1
    return CalcRad((lat1+lat2) / 2) * math.acos(a)

def MeterOffset((lat1, lon1), (lat2, lon2)):
    "Return offset in meters of second arg from first."
    dx = EarthDistance((lat1, lon1), (lat1, lon2))
    dy = EarthDistance((lat1, lon1), (lat2, lon1))
    if lat1 < lat2: dy *= -1
    if lon1 < lon2: dx *= -1
    return (dx, dy)

def isotime(s):
    "Convert timestamps in ISO8661 format to and from Unix time."
    if type(s) == type(1):
        return time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime(s))
    elif type(s) == type(1.0):
        date = int(s)
        msec = s - date
        date = time.strftime("%Y-%m-%dT%H:%M:%S", time.gmtime(s))
        return date + "." + `msec`[2:]
    elif type(s) == type(""):
        if s[-1] == "Z":
            s = s[:-1]
        if "." in s:
            (date, msec) = s.split(".")
        else:
            date = s
            msec = "0"
        # Note: no leap-second correction! 
        return calendar.timegm(time.strptime(date, "%Y-%m-%dT%H:%M:%S")) + float("0." + msec)
    else:
        raise TypeError

if __name__ == '__main__':
    import readline
    print "This is the exerciser for the Python gps interface."
    session = gps()
    session.set_raw_hook(lambda s: sys.stdout.write(s + "\n"))
    try:
        while True:
            commands = raw_input("> ")
            session.query(commands+"\n")
            print session
    except EOFError:
        print "Goodbye!"
    del session

# gps.py ends here
